### PR TITLE
Rollback migrations

### PR DESCRIPTION
**Pull request by aduffeck**
The decomposedfs now supports rolling back migrations (starting with 0004). It also got a Migrations() method which returns the list of migrations incl. their states.

### CODE REVIEW

File: pkg/storage/utils/decomposedfs/migrator/0004_switch_to_messagepack_space_index.go [hunk 1]
```diff
@@ -118,3 +116,82 @@ func migrateSpaceIndex(indexPath, dirIndexPath, cacheKey string) error {
...
+	for _, spaceType := range []string{"personal", "project", "share"} {
+		indexPath := filepath.Join(root, "indexes", "by-type", spaceType+".mpk")
+		dirIndexPath := filepath.Join(root, "indexes", "by-type", spaceType)
+
+		_, err := os.Stat(indexPath)
+		if err != nil || os.IsNotExist(err) {
+			continue
+		}
+
+		m.log.Info().Str("root", m.lu.InternalRoot()).Msg("Migrating " + indexPath + " to messagepack index format...")
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/0004_switch_to_messagepack_space_index.go:164]:
```suggestion
		m.log.Info().Str("root", m.lu.InternalRoot()).Msg("Migrating " + indexPath + " to symlinks index format...")
```

File: pkg/storage/utils/decomposedfs/migrator/0001_create_spaces_directory_structure.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ import (
 )
 
 // Migration0001 creates the spaces directory structure
-func (m *Migrator) Migration0001() (Result, error) {
+func (m *Migrator) Up0001() (Result, error) {
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/0001_create_spaces_directory_structure.go:33]: I like the old naming more. Can we go for `MigrationXXXX` and `RollbackXXXX`?

File: pkg/storage/utils/decomposedfs/migrator/migrator.go [hunk 3]
```diff
@@ -61,6 +64,82 @@ func New(lu *lookup.Lookup, log *zerolog.Logger) Migrator {
...
+	}
+
+	msg := "Running migration " + id + "..."
+	if rollback {
+		msg += " (down)"
+	}
+	m.log.Info().Msg(msg)
+
+	migrateMethod := reflect.ValueOf(m).MethodByName(method)
+	v := migrateMethod.Call(nil)
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/migrator.go:122]: Can we avoid the reflect magic? I'd be more fine with a hardcoded list.

File: pkg/storage/utils/decomposedfs/migrator/0005_fix_messagepack_space_index_format.go [hunk 4]
```diff
@@ -56,13 +56,52 @@ func (m *Migrator) Migration0005() (Result, error) {
...
-			return resultFailed, err
+			return stateFailed, err
 		}
 	}
 	m.log.Info().Msg("done.")
-	return resultSucceeded, nil
+	return stateSucceeded, nil
+}
+
+func (m *Migrator) Down0005() (Result, error) {
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/0005_fix_messagepack_space_index_format.go:70]: exported func `Down0005` should have comment or be unexported

File: pkg/storage/utils/decomposedfs/migrator/migrator.go [hunk 5]
```diff
@@ -32,16 +33,18 @@ import (
...
-	resultSucceededRunAgain = "runagain"
+	statePending           = "pending"
+	stateFailed            = "failed"
+	stateSucceeded         = "succeeded"
+	stateDown              = "down"
+	stateSucceededRunAgain = "runagain"
 )
 
-type migrationState struct {
+type MigrationState struct {
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/migrator.go:43]: exported type `MigrationState` should have comment or be unexported

File: pkg/storage/utils/decomposedfs/migrator/migrator.go [hunk 6]
```diff
@@ -61,6 +64,82 @@ func New(lu *lookup.Lookup, log *zerolog.Logger) Migrator {
 	}
 }
 
+func (m *Migrator) Migrations() (map[string]MigrationState, error) {
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/migrator.go:67]: exported func `Migrations` should have comment or be unexported

File: pkg/storage/utils/decomposedfs/migrator/migrator.go [hunk 7]
```diff
@@ -61,6 +64,82 @@ func New(lu *lookup.Lookup, log *zerolog.Logger) Migrator {
...
+			states[migration] = MigrationState{
+				State: statePending,
+			}
+		}
+	}
+
+	return states, nil
+}
+
+func (m *Migrator) RunMigration(id string, rollback bool) error {
```
Comment (kobergj) [inline pkg/storage/utils/decomposedfs/migrator/migrator.go:87]: exported func `RunMigration` should have comment or be unexported

File: pkg/storage/utils/decomposedfs/migrator/migrator.go (hunk 3 above)
Comment (aduffeck) [reply pkg/storage/utils/decomposedfs/migrator/migrator.go]: @kobergj I came up with https://github.com/cs3org/reva/pull/4083/commits/5076031955383f708b8225b13e75a19c2a155244 which avoids both `reflect` and hard coded lists. wdyt?

**Review by kobergj — APPROVED** [review]
Nice :+1:

**Pull request merged**